Properly check availability of email app

criticalAY opened the pull request:
<!--
- Add the issue number here.

- If you haven't solved the issue completely use "Linked issue #{issue_number}.
- After solving the issue completely change it to "Fixes #{issue_number}.
-->
Fixes #3247

<!-- Add here what changes were made in this issue and if possible provide links. -->
Placed a check for email app and show a toast accordingly

<!-- If possible, please add relevant screenshots / GIFs -->

**Screenshots** 
![Screenshot_58](https://user-images.githubusercontent.com/48384865/221286115-d8fb20ee-b166-406b-af0c-eba266e848be.png)

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpFragment.kt:
```diff
@@ -88,16 +89,21 @@ abstract class HelpFragment : BaseFragment() {
+    val packageManager = requireActivity().packageManager
+    val activities = packageManager.queryIntentActivities(emailIntent, 0)
+    if (activities.isNotEmpty()) {
+      val chooserIntent = Intent.createChooser(emailIntent, "Send Feedback via Email")
```
Since ```chooserIntent``` is used only once, it is recommended to use it as an inline variable instead of declaring it separately in order to avoid writing an extra line of code, like we previously using.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/help/HelpFragment.kt:
```diff
@@ -88,16 +89,21 @@ abstract class HelpFragment : BaseFragment() {
+        ).toUri()
+    }
+    val packageManager = requireActivity().packageManager
+    val activities = packageManager.queryIntentActivities(emailIntent, 0)
```
@criticalAY , ```queryIntentActivities(intent,int)``` is deprecated in android 13. Please use ```queryIntentActivities(intent,ResolveInfoFlags)``` to support android 13.

like 

```
 val activities = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      packageManager.queryIntentActivities(
        emailIntent,
        PackageManager.ResolveInfoFlags.of(MATCH_DEFAULT_ONLY.toLong())
      )
    } else {
      packageManager.queryIntentActivities(
        emailIntent,
        MATCH_DEFAULT_ONLY
      )
    }
```

MohitMaliFtechiz reviewed (changes requested):
@criticalAY , Thanks, Just small changes, Then we are good to go.

criticalAY commented:
@MohitMaliFtechiz I also raised a ticket for the same this is not the right way to handle deprecation. I will come up with a follow-up to handle deprecation properly. For now, I will add the deprecation annotation 
The proper way would be to add a compat file and handle the deprecation there as if we move further we can simply use the new method API just by deleting the deprecated method and making no change to the code. Like here if we surpass the deprecation we will have to make changes to the code itself and from what I am suggesting we will have to make change to only one file and we will be good to go.

MohitMaliFtechiz commented:
@criticalAY , We should not implement deprecated methods here, Placing deprecated code in this PR and then fixing it in another PR is not an ideal approach. Instead, we should fix this method here.

criticalAY commented:
@kelson42 @MohitMaliFtechiz I have introduced a proper way that should be used to handle deprecation this way we won't have to handle deprecation at every single place and can be updated in compat files and used everywhere with ease. Plus I have introduced some other files which will be handy handling other deprecation. 
I have used this idea from AnkiDroid and it is more effective they have this system so the codebase can support even the Old API along with new one. Please take your time to go through it. Thanks!

criticalAY commented:
There are a lot of places that uses "bad form" to handle the deprecation as it is not ideal to declare required api tag and placing a condition here and there

criticalAY commented:
this idea is also proposed in a issue tracker of google and they are working on it https://issuetracker.google.com/issues/243986223

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/CompatV33.kt:
```diff
@@ -0,0 +1,40 @@
+    intent: Intent,
+    flags: ResolveInfoFlagsCompat
+  ): List<ResolveInfo> {
+    return packageManager.queryIntentActivities(
```
Please use an expression body instead of a return statement.

MohitMaliFtechiz replied:
Please use an expression body instead of a return statement.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,153 @@
+ * Flags class that wraps around the bitmask flags used in methods that retrieve package or
+ * application info.
+ */
+open class Flags protected constructor(val value: Long)
```
this too

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,153 @@
+ * Specific flags used for retrieving package info. Example:
+ * `PackageManager.getPackageInfo(packageName, PackageInfoFlags.of(0)`
+ */
+class PackageInfoFlagsCompat private constructor(@PackageInfoFlagsBits value: Long) : Flags(value) {
```
What is the purpose of this class here? It appears to be unused.

criticalAY replied:
that was for `getPackageInfo` I have added and fixed this deprecation as well

MohitMaliFtechiz replied:
@criticalAY, This fix is not part of this ticket. Please only include a fix for this ticket here, and include other parts in a separate PR. We should create a dedicated ticket for a particular deprecation and fix it separately. It'll be easy to track these issues this way.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,153 @@
+  ]
+)
+@Retention(AnnotationRetention.SOURCE)
+annotation class PackageInfoFlagsBits
```
this too.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/ResolveInfoCompat.kt:
```diff
@@ -0,0 +1,164 @@
+  ]
+)
+@Retention(AnnotationRetention.SOURCE)
+annotation class ResolveInfoFlagsBitsCompat
```
Please create this class in a separate file for better code quality. It is essential to keep our code clean and organised, and we should avoid putting multiple classes in a single file, including the signal class.

criticalAY replied:
ok

MohitMaliFtechiz commented:
@criticalAY, This fix is not part of this ticket. Please only include a fix for this ticket here, and include other parts in a separate PR. We should create a dedicated ticket for a particular deprecation and fix it separately. It'll be easy to track these issues this way.

criticalAY commented:
@MohitMaliFtechiz I have removed the unwanted code and I will raise a new ticket once we get this PR in

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,139 @@
+ * Flags class that wraps around the bitmask flags used in methods that retrieve package or
+ * application info.
+ */
+open class Flags protected constructor(val value: Long)
```
Please make it in separate file.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,139 @@
+  ]
+)
+@Retention(AnnotationRetention.SOURCE)
+annotation class PackageInfoFlagsBits
```
this class is unused here, please remove it.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/ResolveInfoCompat.kt:
```diff
@@ -0,0 +1,164 @@
+  ]
+)
+@Retention(AnnotationRetention.SOURCE)
+annotation class ResolveInfoFlagsBitsCompat
```
this class too.

MohitMaliFtechiz replied:
@criticalAY , you forgot to make this class in a separate file.

criticalAY replied:
there is an error if we don't dont declare them in same file

MohitMaliFtechiz replied:
we should fix this.

MohitMaliFtechiz replied:
> i am facing a top level declaration error while committing the code it's not shown in Android studio though

I see now that the class name is ```ResolveInfoFlagsCompat``` and the file name is ```ResolveInfoCompat```. This is why the error is occurring. To resolve it, you can either rename the file or rename the class. Either way, this should fix the issue.

MohitMaliFtechiz reviewed (changes requested):
@criticalAY , I you revert the code as suggested in previous review. please do that changes.

criticalAY commented:
@MohitMaliFtechiz done!

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/compat/PackageManagerCompat.kt:
```diff
@@ -0,0 +1,82 @@
+import android.content.pm.ApplicationInfo
+import android.content.pm.PackageInfo
+import android.content.pm.PackageManager
+import androidx.annotation.LongDef
```
please remove unused imports.

criticalAY commented:
removed unused imports

kelson42 commented:
@criticalAY IMO this check should alos apply if there is an email address in a ZIM content, does it?

criticalAY commented:
I'll look into it once I am back from my trip

criticalAY commented:
@MohitMaliFtechiz @kelson42 I am back from my trip and sorry for the delay but I have made the requested changes now

MohitMaliFtechiz reviewed (approved):
@criticalAY , Thanks
